code/controllers/subsystem/thermal_dissipation.dm

```diff
@@ -32,12 +33,137 @@ var/list/datum/reagents/thermal_dissipation_reagents = list()
 	var/c = currentrun_index
 
 	if(config.thermal_dissipation)
+
 		var/simulate_air = config.reagents_heat_air
+		var/list/turf_air_list = list()
+		var/list/turf_temp_list = list()
+
+		//Variables to reuse
+		var/datum/reagents/R
+		var/turf/T
+		var/datum/gas_mixture/the_air
+		var/emission_factor
+		var/atom/this_potentially_insulative_layer
+		var/i
+		var/air_thermal_mass
+		var/Tr
+		var/Ta
+		var/reagents_thermal_mass_reciprocal
+		var/air_thermal_mass_reciprocal
+		var/slices
+		var/this_slice_energy
+
 		while (c)
 
-			currentrun[c]?.handle_thermal_dissipation(simulate_air)
+			R = currentrun[c]
 			c--
 
+			if(!R)
+				goto tick_check
+
+			//Exchange heat between reagents and the surrounding air.
+			//Although the heat is exchanged directly between reagents and air, for now this is based on thermal radiation, not convection per se.
+
+			if (!(T = get_turf(R.my_atom)))
+				goto tick_check
+
+			if (!(abs((Ta := (turf_temp_list[T] || (turf_temp_list[T] := T.air_temperature()))) - R.chem_temp) >= MINIMUM_TEMPERATURE_DELTA_TO_CONSIDER))  //Do it this way to catch NaNs.
+				goto tick_check
+
+			if (!R.total_volume || !R.total_thermal_mass || R.gcDestroyed || R.my_atom.gcDestroyed || R.my_atom.timestopped)
+				goto tick_check
+
+			//We treat the reagents like a spherical grey body with an emissivity of THERM_DISS_SCALING_FACTOR.
+
+			emission_factor = THERM_DISS_SCALING_FACTOR * (SS_WAIT_THERM_DISS / (1 SECONDS)) * STEFAN_BOLTZMANN_CONSTANT * (36 * PI) ** (1/3) * (CC_PER_U / 1000) ** (2/3) * R.total_volume ** (2/3)
+
+			//Here we reduce thermal transfer to account for insulation of the container.
+			//We iterate though each loc until the loc is the turf T, to account for things like nested containers, each time multiplying emission_factor by a factor than can range between [0 and 1], representing heat insulation.
+
+			this_potentially_insulative_layer = R.my_atom
+			i = ARBITRARILY_LARGE_NUMBER
+			while (emission_factor)
+				emission_factor *= this_potentially_insulative_layer.get_heat_conductivity()
+				if (this_potentially_insulative_layer == T)
+					break
+				else if (i && (this_potentially_insulative_layer := this_potentially_insulative_layer.loc))
+					i--
+				else
+					log_admin("Something went wrong with [R.my_atom]'s handle_heat_dissipation() at iteration #[i] at [this_potentially_insulative_layer].")
+					break //Avoid infinite loops and nullspace issues.
+
+			if (emission_factor)
+
+				Tr = R.chem_temp
+
+				if (max(Tr, Ta) <= THERM_DISS_MAX_SAFE_TEMP)
+
+					reagents_thermal_mass_reciprocal = (1 / R.total_thermal_mass)
+
+					if (simulate_air && !istype(the_air := (turf_air_list[T] || (turf_air_list[T] := T.return_air())), /datum/gas_mixture/unsimulated))
```

Comment:
Out of curiosity, what does the `:=` operator do in this instance? 
My current understanding was that it "_is intended mainly for situations where you've overloaded the operator_"  (to quote the [DM reference](https://secure.byond.com/docs/ref/#/operator/:=)) so you can add extra logic on top of the usual assignment, but I'm not aware of any such overloads. 
Hell, I'm not even sure what happens if you use it without overloading the operator first; my assumption before this was that it defaults to `=` behavior, is there more to it?

Reply:
It does the same as = for an already-declared variable but the byond compiler puts warnings sometimes if you try to use just =, thinking you meant to do ==
I also wanted to be explicit that it wasn't supposed to be == to others reading it in the future

Reply:
It seems there's a performance benefit in that you can use the value directly rather than writing it to the list *then* re reading it out of the list in the next list

check this http://www.byond.com/forum/post/1225436

Reply:
Consider the `||=` operator for this

Reply:
> Why not write
```
if (turf_air_list[T])
    the_air = turf_air_list[T]
```
in that case you have to do two list reads if turf_air_list contains T, whereas with this way you only have to do one list read